File: modules/desktop-client-sdk/src/main/java/com/avandocmsg/messenger/desktop/sdk/call/InProcessCallClient.java
package com.avandocmsg.messenger.desktop.sdk.call;

import com.avandocmsg.messenger.desktop.sdk.api.KorusApiClient;
import com.avandocmsg.messenger.desktop.sdk.model.CallJoinResponse;
import com.avandocmsg.messenger.desktop.sdk.model.CallSignalResponse;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;
import java.util.function.Supplier;

public final class InProcessCallClient implements AutoCloseable {

    private final KorusApiClient api;
    private final Supplier<CallAudioMedia> mediaFactory;
    private final AtomicBoolean leaving = new AtomicBoolean();
    private volatile Runnable hangupListener = () -> {};
    private CallJoinResponse join;
    private String token;
    private CallAudioMedia media;
    private Thread poller;

    public InProcessCallClient(KorusApiClient api) {
        this(api, NativeCallAudioMedia::new);
    }

    public InProcessCallClient(KorusApiClient api, Supplier<CallAudioMedia> mediaFactory) {
        this.api = Objects.requireNonNull(api, "api");
        this.mediaFactory = Objects.requireNonNull(mediaFactory, "mediaFactory");
    }

    public CallJoinResponse start(String token, String chatId, String kind, String mediaIntent)
        throws Exception {
        this.token = required(token, "token");
        join = api.createCall(this.token, chatId, kind, mediaIntent);
        return connectMedia();
    }

    public CallJoinResponse join(String token, String chatId, String sessionId) throws Exception {
        this.token = required(token, "token");
        join = api.joinCall(this.token, chatId, sessionId);
        return connectMedia();
    }

    public CallJoinResponse join() {
        return join;
    }

    public boolean mediaReady() {
        return media != null && media.mediaReady();
    }

    public void sendPcmu(byte[] payload) {
        if (media != null) {
            media.sendPcmu(payload);
        }
    }

    public void onPcmu(Consumer<byte[]> listener) {
        if (media != null) {
            media.onPcmu(listener);
        }
    }

    public void onHangup(Runnable listener) {
        hangupListener = listener == null ? () -> {} : listener;
    }

    public void leave() {
        if (!leaving.compareAndSet(false, true)) {
            return;
        }
        if (poller != null) {
            poller.interrupt();
        }
        if (media != null) {
            media.close();
        }
        if (join != null && token != null) {
            try {
                api.leaveCall(token, join);
            } catch (RuntimeException ignored) {
                // The local media path is already closed.
            }
        }
        hangupListener.run();
    }

    @Override
    public void close() {
        leave();
    }

    private CallJoinResponse connectMedia() throws Exception {
        media = mediaFactory.get();
        api.sendCallSignal(token, join, "offer", media.createOffer(), null);
        var answer = awaitAnswer();
        media.connect(answer.sdp());
        poller = Thread.ofVirtual().name("korus-desktop-call-signals").start(this::pollUntilHangup);
        return join;
    }

    private CallSignalResponse awaitAnswer() throws InterruptedException {
        var deadline = System.nanoTime() + 20_000_000_000L;
        while (System.nanoTime() < deadline && !leaving.get()) {
            var signals = api.pollCallSignals(token, join);
            for (var signal : signals) {
                if ("answer".equals(signal.type()) && signal.sdp() != null) {
                    return signal;
                }
                if ("error".equals(signal.type())) {
                    throw new IllegalStateException(
                        signal.errorCode() == null ? "call media rejected" : signal.errorCode()
                    );
                }
            }
            Thread.sleep(200);
        }
        throw new IllegalStateException("call answer timed out");
    }

    private void pollUntilHangup() {
        while (!leaving.get() && !Thread.currentThread().isInterrupted()) {
            try {
                var signals = api.pollCallSignals(token, join);
                for (var signal : signals) {
                    if ("hangup".equals(signal.type()) || "session_ended".equals(signal.type())) {
                        leave();
                        return;
                    }
                }
                Thread.sleep(400);
            } catch (RuntimeException | InterruptedException stopped) {
                return;
            }
        }
    }

    private static String required(String value, String name) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException(name + " required");
        }
        return value;
    }
}

File: mobile/mobile-client-sdk/src/commonMain/kotlin/com/avandocmsg/messenger/mobile/sdk/call/InProcessCallClient.kt
package com.avandocmsg.messenger.mobile.sdk.call

import com.avandocmsg.messenger.mobile.sdk.api.KorusApiClient
import com.avandocmsg.messenger.mobile.sdk.model.CallJoinDto
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.util.concurrent.atomic.AtomicBoolean

class InProcessCallClient(
    private val api: KorusApiClient,
    private val scope: CoroutineScope,
    private val mediaFactory: () -> CallAudioMedia
) : AutoCloseable {
    private val leaving = AtomicBoolean(false)
    private var token: String? = null
    private var join: CallJoinDto? = null
    private var media: CallAudioMedia? = null
    private var pollJob: Job? = null
    private var hangupListener: () -> Unit = {}

    val activeJoin: CallJoinDto?
        get() = join

    suspend fun start(
        token: String,
        chatId: String,
        kind: String = "group",
        mediaIntent: String = "audio"
    ): CallJoinDto {
        this.token = required(token, "token")
        join = api.createCall(this.token!!, chatId, kind, mediaIntent)
        connectMedia()
        return join!!
    }

    suspend fun join(token: String, chatId: String, sessionId: String): CallJoinDto {
        this.token = required(token, "token")
        join = api.joinCall(this.token!!, chatId, sessionId)
        connectMedia()
        return join!!
    }

    fun mediaReady(): Boolean = media?.mediaReady() == true

    fun sendPcmu(payload: ByteArray) {
        media?.sendPcmu(payload)
    }

    fun onPcmu(listener: (ByteArray) -> Unit) {
        media?.onPcmu(listener)
    }

    fun onHangup(listener: () -> Unit) {
        hangupListener = listener
    }

    fun leave() {
        if (!leaving.compareAndSet(false, true)) {
            return
        }
        pollJob?.cancel()
        media?.close()
        media = null
        val activeJoin = join
        val activeToken = token
        if (activeJoin != null && activeToken != null) {
            scope.launch {
                runCatching { api.leaveCall(activeToken, activeJoin) }
            }
        }
        join = null
        hangupListener.invoke()
    }

    override fun close() {
        leave()
    }

    private suspend fun connectMedia() {
        val activeToken = required(token, "token")
        val activeJoin = join ?: throw IllegalStateException("join missing")
        media = mediaFactory()
        api.sendCallSignal(activeToken, activeJoin, type = "offer", sdp = media!!.createOffer())
        val answer = awaitAnswer(activeToken, activeJoin)
        media!!.connect(answer)
        pollJob = scope.launch { pollUntilHangup(activeToken, activeJoin) }
    }

    private suspend fun awaitAnswer(token: String, join: CallJoinDto): String {
        val deadline = System.nanoTime() + 20_000_000_000L
        while (System.nanoTime() < deadline && !leaving.get()) {
            for (signal in api.pollCallSignals(token, join)) {
                when (signal.type) {
                    "answer" -> signal.sdp?.let { return it }
                    "error" -> throw IllegalStateException(signal.errorCode ?: "call media rejected")
                }
            }
            delay(200)
        }
        throw IllegalStateException("call answer timed out")
    }

    private suspend fun pollUntilHangup(token: String, join: CallJoinDto) {
        while (!leaving.get() && scope.isActive) {
            try {
                for (signal in api.pollCallSignals(token, join)) {
                    if (signal.type == "hangup" || signal.type == "session_ended") {
                        leave()
                        return
                    }
                }
                delay(400)
            } catch (_: Exception) {
                return
            }
        }
    }

    private fun required(value: String?, name: String): String {
        require(!value.isNullOrBlank()) { "$name required" }
        return value
    }
}
